lms/services/grouping.py

```diff
@@ -26,74 +26,48 @@ def upsert(self, grouping):
 
         return db_grouping or grouping
 
-    def upsert_canvas_section(
-        self, tool_consumer_instance_guid, context_id, section_id, section_name
+    @staticmethod
+    def generate_authority_provided_id(
+        tool_consumer_instance_guid, lms_id, parent: Grouping, type_: Grouping.Type
     ):
-        """
-        Upsert a Grouping for a course section.
-
-        :param tool_consumer_instance_guid: Tool consumer GUID
-        :param context_id: Course id the section is a part of
-        :param section_id: A section id for a section group
-        :param section_name: The name of the section
-        """
-
-        section_authority_provided_id = hashed_id(
-            tool_consumer_instance_guid, context_id, section_id
-        )
-
-        course = self._course_service.get(
-            self._course_service.generate_authority_provided_id(
-                tool_consumer_instance_guid, context_id
+        if type_ == Grouping.Type.CANVAS_SECTION:
+            return hashed_id(tool_consumer_instance_guid, parent.lms_id, lms_id)
+        if type_ == Grouping.Type.COURSE:
+            return CourseService.generate_authority_provided_id(
+                tool_consumer_instance_guid, lms_id
             )
-        )
 
-        return self.upsert(
-            CanvasSection(
-                application_instance=self._application_instance,
-                authority_provided_id=section_authority_provided_id,
-                lms_id=section_id,
-                lms_name=section_name,
-                parent_id=course.id,
-            )
-        )
+        return hashed_id(tool_consumer_instance_guid, parent.lms_id, type_, lms_id)
 
-    def upsert_canvas_group(  # pylint: disable=too-many-arguments
+    def upsert_with_parent(  # pylint: disable=too-many-arguments
```

Comment:
Not sure about this name. 

I don't want to merge this with `upsert` as I prefer the service clients don't have to bother constructing Grouping objects themselves.

Reply:
It looks to me like nothing outside of this service actually calls the `upsert()` method so I think you could make it a private `_upsert()` method and move it to the bottom of the class, then you could name the public method `upsert()`:

```python
class GroupingService:
    ...

    def upsert(  # pylint: disable=too-many-arguments
        self,
        tool_consumer_instance_guid,
        lms_id,
        lms_name,
        parent: Grouping,
        type_: Grouping.Type,
        extra=None,
    ):
        return self._upsert(...)

    def _upsert(self, grouping):
        ...
``` 

This would mean that the `_upsert()` method would have to be unit-tested indirectly via `upsert()` (to avoid the tests calling a private method) but that's probably a good thing since `_upsert()` is really just an implementation details of `upsert()`

Reply:
I wonder about the `# pylint: disable=too-many-arguments` though. Why don't you want the caller to have to construct `Grouping` objects? Instead of:

```python
grouping_service.upsert(
    tool_consumer_instance_guid=tool_consumer_instance_guid,
    lms_id=lms_id,
    lms_name=lms_name,
    parent=parent_grouping,
    type_=Grouping.Type.*,
    extra={whatever},
)
```

It'd be:

```python
grouping_service.upsert(
    Grouping(
        tool_consumer_instance_guid=tool_consumer_instance_guid,
        lms_id=lms_id,
        lms_name=lms_name,
        parent=parent_grouping,
        type_=Grouping.Type.*,
        extra={whatever},
    )
)
```

It shouldn't be _too_ much harder from the point of view of the caller and it avoids the `too-many-arguments`.

`too-many-arguments` is a common issue with services when the caller passes in all the arguments directly and the service constructs the model object. And I think having the caller pass in the model object is probably the solution.

Having said that, it'd probably take some more work to make passing-in-the-model work as nicely as it does in my example above, so I think just leaving things as they are is probably the most pragmatic solution right now.

Reply:
Made _upsert private :+1: 

I don't think it has many practical implications here but I rather the rest of the code not construct ORM objects outside the service so that becomes the responsibility of the service, if the views construct the object they might as well `db_session.add()` them.

To be fair, in this case the service it's not doing that much  (calling _upsert, generating the authority provided id)

Reply:
> I think we can rename upsert_with_parent() to just upsert() now?

I  think if the public method was called `upsert` I would expect it to work with groupings that don't have parents (courses). I took the route of keeping the `_with_parent` naming but removing support for when parent==None which added a few branches in the code.